File: server/ai/rag.ts
import { ulid } from 'ulid';
import { eq } from 'drizzle-orm';
import { db, schema, addEmbedding, removeEmbeddingsBySource, searchEmbeddings } from '../db/index.js';
import { generateEmbeddings, generateEmbedding, isEmbeddingsConfigured } from './embeddings.js';

// Sync a single journal entry to embeddings
export async function syncJournalEmbeddings(journalId: string): Promise<number> {
  const configured = await isEmbeddingsConfigured();
  if (!configured) {
    console.warn('Embeddings not configured, skipping sync');
    return 0;
  }

  // Get the journal entry
  const [journal] = await db
    .select()
    .from(schema.journalEntries)
    .where(eq(schema.journalEntries.id, journalId))
    .limit(1);

  if (!journal) {
    throw new Error('Journal entry not found');
  }

  // Remove old embeddings for this entry
  removeEmbeddingsBySource('journal', journalId);

  // Generate new embeddings
  const content = journal.title ? `${journal.title}\n\n${journal.content}` : journal.content;
  const results = await generateEmbeddings(content);

  // Store embeddings
  for (const result of results) {
    const id = ulid();
    addEmbedding(id, result.embedding, {
      sourceType: 'journal',
      sourceId: journalId,
      chunkIndex: result.chunkIndex,
      chunkText: result.chunkText,
    });
  }

  return results.length;
}

// Sync all journal entries (for initial setup or rebuild)
export async function syncAllJournalEmbeddings(
  onProgress?: (current: number, total: number) => void
): Promise<{ synced: number; chunks: number }> {
  const configured = await isEmbeddingsConfigured();
  if (!configured) {
    throw new Error('Embeddings not configured');
  }

  const journals = await db.select().from(schema.journalEntries);
  let totalChunks = 0;

  for (let i = 0; i < journals.length; i++) {
    const journal = journals[i];
    try {
      const chunks = await syncJournalEmbeddings(journal.id);
      totalChunks += chunks;
      onProgress?.(i + 1, journals.length);
    } catch (error) {
      console.error(`Failed to sync journal ${journal.id}:`, error);
    }
  }

  return { synced: journals.length, chunks: totalChunks };
}

// Delete embeddings when a journal entry is deleted
export function deleteJournalEmbeddings(journalId: string): void {
  removeEmbeddingsBySource('journal', journalId);
}

// Semantic search for relevant context
export interface RAGSearchResult {
  sourceType: string;
  sourceId: string;
  content: string;
  score: number;
  metadata?: {
    title?: string;
    createdAt?: Date;
  };
}

export async function searchForContext(
  query: string,
  limit: number = 5
): Promise<RAGSearchResult[]> {
  const configured = await isEmbeddingsConfigured();
  if (!configured) {
    return [];
  }

  // Generate embedding for the query
  const queryEmbedding = await generateEmbedding(query);

  // Search for similar embeddings
  const results = searchEmbeddings(queryEmbedding, limit * 2); // Get more to dedupe

  // Deduplicate by source and get full content
  const sourceMap = new Map<string, RAGSearchResult>();

  for (const result of results) {
    const key = `${result.sourceType}:${result.sourceId}`;
    if (sourceMap.has(key)) continue;

    // Get full source content
    if (result.sourceType === 'journal') {
      const [journal] = await db
        .select()
        .from(schema.journalEntries)
        .where(eq(schema.journalEntries.id, result.sourceId))
        .limit(1);

      if (journal) {
        sourceMap.set(key, {
          sourceType: result.sourceType,
          sourceId: result.sourceId,
          content: journal.content,
          score: 1 - result.distance, // Convert distance to similarity score
          metadata: {
            title: journal.title || undefined,
            createdAt: journal.createdAt,
          },
        });
      }
    }

    if (sourceMap.size >= limit) break;
  }

  return Array.from(sourceMap.values());
}

// Format RAG results as context for the AI
export function formatRAGContext(results: RAGSearchResult[]): string {
  if (results.length === 0) {
    return '';
  }

  const sections = results.map((result, index) => {
    const header = result.metadata?.title
      ? `[${result.metadata.title}]`
      : `[Entry from ${result.metadata?.createdAt?.toLocaleDateString() || 'unknown date'}]`;

    return `--- Context ${index + 1} ${header} ---\n${result.content}`;
  });

  return `Here is relevant information from the user's personal knowledge base:\n\n${sections.join('\n\n')}`;
}

// Check if embeddings exist for a source
export async function hasEmbeddings(sourceType: string, sourceId: string): Promise<boolean> {
  // This is a simple check - in production you'd query the metadata table
  try {
    const results = searchEmbeddings(new Array(1536).fill(0), 1, sourceType);
    return results.some((r) => r.sourceId === sourceId);
  } catch {
    return false;
  }
}

// Get embedding stats
export interface EmbeddingStats {
  totalJournals: number;
  indexedJournals: number;
  totalChunks: number;
  isConfigured: boolean;
}

export async function getEmbeddingStats(): Promise<EmbeddingStats> {
  const configured = await isEmbeddingsConfigured();
  const totalJournals = (await db.select().from(schema.journalEntries)).length;

  if (!configured) {
    return {
      totalJournals,
      indexedJournals: 0,
      totalChunks: 0,
      isConfigured: false,
    };
  }

  // Count unique indexed journals - we'll estimate from chunks
  // In production, query the metadata table directly
  return {
    totalJournals,
    indexedJournals: 0, // Would need metadata query
    totalChunks: 0, // Would need metadata query
    isConfigured: true,
  };
}

File: backend/ai/rag.py
import struct
import time

from backend.db.connection import get_db, row_to_dict
from backend.ai.embeddings import generate_embedding, generate_embeddings, is_embeddings_configured


def _insert_embedding(id: str, embedding: list[float], source_type: str, source_id: str, chunk_index: int, chunk_text: str) -> None:
    db = get_db()
    blob = struct.pack(f'{len(embedding)}f', *embedding)
    db.execute(
        'INSERT OR REPLACE INTO vec_embeddings(id, embedding) VALUES (?, ?)',
        (id, blob),
    )
    db.execute(
        'INSERT OR REPLACE INTO embedding_metadata(id, source_type, source_id, chunk_index, chunk_text, created_at) VALUES (?, ?, ?, ?, ?, ?)',
        (id, source_type, source_id, chunk_index, chunk_text, int(time.time())),
    )
    db.commit()


def _delete_embeddings_by_source(source_type: str, source_id: str) -> None:
    db = get_db()
    ids = [r['id'] for r in db.execute(
        'SELECT id FROM embedding_metadata WHERE source_type=? AND source_id=?',
        (source_type, source_id),
    ).fetchall()]
    for id in ids:
        db.execute('DELETE FROM vec_embeddings WHERE id=?', (id,))
        db.execute('DELETE FROM embedding_metadata WHERE id=?', (id,))
    db.commit()


def _search_similar(query_embedding: list[float], limit: int = 5, source_type: str | None = None) -> list[dict]:
    db = get_db()
    blob = struct.pack(f'{len(query_embedding)}f', *query_embedding)
    sql = """
        SELECT v.id, v.distance, m.source_type, m.source_id, m.chunk_index, m.chunk_text
        FROM vec_embeddings v
        JOIN embedding_metadata m ON v.id = m.id
        WHERE v.embedding MATCH ?
    """
    params: list = [blob]
    if source_type:
        sql += ' AND m.source_type = ?'
        params.append(source_type)
    sql += ' ORDER BY v.distance LIMIT ?'
    params.append(limit)
    return [dict(r) for r in db.execute(sql, params).fetchall()]


def sync_journal_embeddings(journal_id: str) -> int:
    if not is_embeddings_configured():
        return 0
    db = get_db()
    row = db.execute('SELECT * FROM journal_entries WHERE id=?', (journal_id,)).fetchone()
    if not row:
        raise ValueError('Journal entry not found')
    _delete_embeddings_by_source('journal', journal_id)
    content = f"{row['title']}\n\n{row['content']}" if row['title'] else row['content']
    results = generate_embeddings(content)
    from ulid import ULID
    for r in results:
        _insert_embedding(str(ULID()), r['embedding'], 'journal', journal_id, r['chunk_index'], r['chunk_text'])
    return len(results)


def sync_all_journal_embeddings() -> dict:
    if not is_embeddings_configured():
        raise ValueError('Embeddings not configured')
    db = get_db()
    journals = db.execute('SELECT id FROM journal_entries').fetchall()
    total_chunks = 0
    for j in journals:
        try:
            total_chunks += sync_journal_embeddings(j['id'])
        except Exception as e:
            print(f'Failed to sync journal {j["id"]}: {e}')
    return {'synced': len(journals), 'chunks': total_chunks}


def delete_journal_embeddings(journal_id: str) -> None:
    _delete_embeddings_by_source('journal', journal_id)


def search_for_context(query: str, limit: int = 5) -> list[dict]:
    if not is_embeddings_configured():
        return []
    try:
        query_embedding = generate_embedding(query)
        results = _search_similar(query_embedding, limit * 2)
    except Exception:
        return []
    db = get_db()
    seen: dict[str, dict] = {}
    for r in results:
        key = f"{r['source_type']}:{r['source_id']}"
        if key in seen:
            continue
        if r['source_type'] == 'journal':
            row = db.execute('SELECT * FROM journal_entries WHERE id=?', (r['source_id'],)).fetchone()
            if row:
                seen[key] = {
                    'sourceType': r['source_type'],
                    'sourceId': r['source_id'],
                    'content': row['content'],
                    'score': 1 - r['distance'],
                    'metadata': {
                        'title': row['title'],
                        'createdAt': row['created_at'],
                    },
                }
        if len(seen) >= limit:
            break
    return list(seen.values())


def format_rag_context(results: list[dict]) -> str:
    if not results:
        return ''
    sections = []
    for i, r in enumerate(results):
        title = r.get('metadata', {}).get('title') or 'Entry'
        sections.append(f'--- Context {i + 1} [{title}] ---\n{r["content"]}')
    return 'Here is relevant information from the user\'s personal knowledge base:\n\n' + '\n\n'.join(sections)


def get_embedding_stats() -> dict:
    db = get_db()
    total = db.execute('SELECT COUNT(*) FROM journal_entries').fetchone()[0]
    configured = is_embeddings_configured()
    return {'totalJournals': total, 'indexedJournals': 0, 'totalChunks': 0, 'isConfigured': configured}
